### PR TITLE
Instance annotations for recursive hierarchies

### PR DESCRIPTION
**Pull request by HeikoTheissen**

### CODE REVIEW

**Comment by HeikoTheissen** [comment]
To be merged after Albatross 2023-06-27

**Review by ThomasChadzelek — APPROVED** [review]

**Review by uhlmannm — APPROVED** [review]

File: vocabularies/Hierarchy.json [hunk 1]
```diff
@@ -26,66 +26,64 @@
...
...vels(..., NodeProperty='ID', Levels=2)\n&$top=10\n```\n(where `...,` stands for hierarchy nodes and hierarchy qualifier)\nthe following collections of hierarchy nodes are distinguished:\n|Collection|Definition|Value|Where in request|\n|----------|----------|-----|----------------|\n|sub-hierarchy|output set of a `descendants` transformation, possibly embedded in a `rolluprecursive` transformation, that is not preceded by an `ancestors` or `descendants` transformation|US sales organizations|rows 1–2|\n|matching nodes|see [`MatchCount`](#MatchCount)|US sales organizations with \"New York\" in their name|output set of `filter` transformation in row 3|\n|unlimited hierarchy|output set of the last `ancestors`, `descendants` or `traverse` transformation, possibly embedded in a `rolluprecursive` transformation, disregarding numeric fifth parameters|US sales organizations with leaves containing \"New York\"|rows 1–3|\n|limited hierarchy|output set of the last `ancestors`, `descendants`, `traverse` or [`Hierarchy.TopLevels`](#TopLevels) transformation, possibly embedded in a `rolluprecursive` transformation|2 levels of US sales organizations with leaves containing \"New York\"|rows 1–4|\n",
```
Comment (agoerler) [inline vocabularies/Hierarchy.json:35]: missing closing parenthesis: `filter(contains(Name, 'New York')` -> `filter(contains(Name, 'New York'))`

**Review by agoerler — CHANGES_REQUESTED** [review]
@HeikoTheissen - found out too late that `*.md` and `*.json` are generated from *.xml. Hence my comments are on the `*.json` file.

File: vocabularies/Hierarchy.json [hunk 2]
```diff
@@ -26,66 +26,64 @@
...
-        "@Core.LongDescription": "If a `NodeTypeProperty` exists, the external key is unique only in combination with it.\n            Or the external key can coincide with the [`Aggregation.RecursiveHierarchy/NodeProperty`](https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Aggregation.V1.html#RecursiveHierarchyType)."
+        "@Core.Description": "Human-readable key value for a node",
+        "@Core.LongDescription": "If a `NodeType` exists, the external key is unique only in combination with it.\n            Or the external key can coincide with the [`Aggregation.RecursiveHierarchy/NodeProperty`](https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Aggregation.V1.html#RecursiveHierarchyType)."
       },
-      "NodeTypeProperty": {
-        "$Type": "Edm.PropertyPath",
+      "NodeType": {
         "$Nullable": true,
-        "@Core.Description": "String property holding the type of a node",
+        "@Core.Description": "Type of a node",
```
[the remark this answers is not in the document]
Comment (HeikoTheissen) [reply vocabularies/Hierarchy.json:43]: `Edm.String` is the default value for `$Type` in CSDL JSON, thus can be omitted.

**Review by agoerler — APPROVED** [review]

**Comment by HeikoTheissen** [comment]
Shall we include an (SAP-specific) annotation for cycles here? Because the OASIS vocabulary does not contain one.

**Comment by ThomasChadzelek** [comment]
> Shall we include an (SAP-specific) annotation for cycles here? Because the OASIS vocabulary does not contain one.

Cycles? I was always under the impression that a tree is a special case of a directed **acyclic** graph ;-)

**Comment by uhlmannm** [comment]
> > Cycles? I was always under the impression that a tree is a special case of a directed **acyclic** graph ;-)
> 
> Not at SAP. image

Would it make sense to discuss this case when it comes up? For me it is unclear what such a hierarchy would mean. The node for "Anne" is in the tree twice (somewhat like a multiple parents case) but its children have to be included only in one of the places.

**Pull request merged**